source/fundamentals/logging.txt

```diff
@@ -0,0 +1,378 @@
+
+.. tabs::
+
+   .. tab:: Logback
+      :tabid: Logback-binding
+
+      Select the build tool you are using in your project.
+
+      .. tabs::
+
+         .. tab:: Maven
+            :tabid: maven
+                        
+            Add the following dependency to your ``pom.xml`` file.
+   
+            .. code-block:: xml
+   
+               <dependency>
+                 <groupId>ch.qos.logback</groupId>
+                 <artifactId>logback-classic</artifactId>
+                 <version>1.2.3</version>
+               </dependency>
+
+         .. tab:: Gradle
+            :tabid: gradle
+   
+            Add the following dependency to your ``build.gradle`` file.
+   
+            .. code-block:: none
+   
+               dependencies {
+                  implementation 'ch.qos.logback:logback-classic:1.2.3'
+               }
+
+      Once you have installed the above dependency, connect to your
+      MongoDB instance and retrieve a document with the following code.
+
+      .. include:: /includes/fundamentals/logging.rst
+
+      You should see output that resembles the following:
+
+      .. code-block:: none
+
+         ...
+         1156 [main] DEBUG o.m.driver.protocol.command - Sending command '{"find": "<your collection>", "filter": {}, "limit": 1, "singleBatch": true, "$db": "<your db>", "lsid": {"id": {"$binary": {"base64": "<server session>", "subType": "04"}}}}' with request id 13 to database <your db> on connection <your connection> to server <your server>
+         1173 [main] DEBUG o.m.driver.protocol.command - Execution of command with request id 13 completed successfully in 30.01 ms on connection <your connection> to server <your server>
+   
+      For more information on Logback, see the
+      `Logback manual <http://logback.qos.ch/manual/>`__.
+   
+   .. tab:: Log4j2
+      :tabid: Log4j2-binding
+
+      Select the build tool you are using in your project.
+      
+      .. tabs::
+
+         .. tab:: Maven
+            :tabid: maven
+                        
+            Add the following dependency to your ``pom.xml`` file.
+   
+            .. code-block:: xml
+   
+               <dependency>
+                 <groupId>org.apache.logging.log4j</groupId>
+                 <artifactId>log4j-slf4j-impl</artifactId>
+                 <version>2.14.1</version>
+               </dependency>
+
+         .. tab:: Gradle
+            :tabid: gradle
+   
+            Add the following dependency to your ``build.gradle`` file.
+   
+            .. code-block:: none
+   
+               dependencies {
+                  implementation 'org.apache.logging.log4j:log4j-slf4j-impl:2.14.1'
+               }
+    
+      Once you have installed the above dependency, log an error using the
+      following code.
+
+      .. code-block:: java
+
+         import org.slf4j.Logger;
+         import org.slf4j.LoggerFactory;
+
+         ...
+
+         Logger logger = LoggerFactory.getLogger("MyApp");
```

Comment:
This is inconsistent with the Logback example, which shows actual driver usage.  I think the intention was to show this simple sort of example here, and show the MongoClient example in the next section, but I'm not sure.

Reply:
Thanks for the note. The reason for the difference between the examples for Logback and Log4J2 is because Log4J2 has a default logger level of "[ERROR](https://logging.apache.org/log4j/2.x/manual/configuration.html)", where as Logback has a default logger level of "[DEBUG](https://github.com/qos-ch/logback/blob/master/logback-examples/src/main/resources/chapters/configuration/sample0.xml)" . This means that performing basic operations with the MongoDB Java driver will give logger output for Logback but not for Log4j2. 

There is a way to change the root logger level of Log4J2 [programmatically](https://logging.apache.org/log4j/log4j-2.4/faq.html#reconfig_level_from_code), but it is not supported by the public api. Configuring your logging framework to change the log level through a config file seemed like a separate task, and is the subject of the next section, so I didn't want to introduce it here. The most natural solution to this problem seemed to me to use different examples, as you can only use one log framework at a time. 

I had trouble figuring out how to log an error with the MongoDB Java driver using easy to understand code, so I went with this solution (also discussed in this comment https://github.com/mongodb/docs-java/pull/76#discussion_r618889732). If there is a relatively straightforward way to have the driver log an error, we could swap this or both examples to use that code, but I am not sure whether or not it would improve the experience for learners.

Please let me know if it would have been appropriate for me to have reached out to you directly in regards to the question of "is there a way to have the MongoDB Java driver easily log an error," or if waiting to have this surface in review was the right approach.

Reply:
There's isn't an easy way to force the driver to log an error, as in most case the driver instead just throws an exception.
Given the difference between Logback and Log4J, I don't see an easy way forward. It does seem strange though that there is only one place where we show actual driver-generated log messages, and we only do it for one of the two logging implementations.

Here's one alternative: Do it the log4J way for Logback as well, and add a complete example at the end for each logging implementation, showing how to set logging level for one of the actual loggers used the driver, and a sample message generated by the driver for that level.   It seems to me not going much further than you already have to add configuration of loggers for the driver in the configuration files for each example.

Happy to discuss it over Slack

Reply:
Slack conversation my takeaways message: "Sounds good, I'll add a mention of the default log level to help readers understand the difference between the examples.
I didn't realize that support required frequent logger-name configuration, but that makes sense. As it is a common use case for clients, I agree it makes sense to include the example explicitly rather than linking to external docs. I'll add the example (hopefully by the end of the day) and ping you on slack when it is ready (let me know if you would rather I ping you on a different channel).